libs/media-player/src/lib/timeline-ruler.tsx

```diff
@@ -0,0 +1,154 @@
+    setRulerWidth,
+}: TimelineProps): JSX.Element => {
+    const roundedSecondsInDuration: number = Math.ceil(duration);
+
+    const {
+        rulerTickXCoordinateOffset,
+        rulerTickFrequencyInSeconds,
+        timecodeDisplayFrequencyInSeconds,
+    } = zoomLevelConfig;
+
+    /**
+     * Create an array of all the ticks on the ruler from the seconds in the media duration
+     * based on the zoom level config e.g.,
+     * {
+     *      zoomLevel: 0,
+     *      rulerTickXCoordinateOffset: 40,  // 40px between ticks at this zoom level
+     *      rulerTickFrequencyInSeconds: 1,  // a tick for each second, or less as the zoom decreases
+     *      timecodeDisplayFrequencyInSeconds: 10  // display the timecode every 10th second in the timeline
+     * }
+     */
+    const ticksOnTimelineAtCurrentZoom = [...Array(roundedSecondsInDuration).keys()]
+        .map((second) => {
+            if (second % rulerTickFrequencyInSeconds === 0) {
+                if (second % timecodeDisplayFrequencyInSeconds === 0) {
+                    const timeCode = asFormattedMediaTimecodeString(second);
+
+                    return {
+                        second: second,
+                        label: timeCode,
+                    };
+                }
+
+                return {
+                    second: second,
+                };
+            }
+
+            return null;
+        })
+        .filter((second) => !isNull(second)) as unknown as TimelineTimecodeProps[];
+
+    // Set up sequential timeline canvases
+
+    const rulerUnitWidthInPixels = rulerTickXCoordinateOffset + RULER_TICK_WIDTH_IN_PIXELS;
+
+    const rulerWidthInPixels = rulerUnitWidthInPixels * ticksOnTimelineAtCurrentZoom.length;
+
+    // Pass ruler width to parent Timeline component for playhead tracking
+    useEffect(() => {
+        setRulerWidth(rulerWidthInPixels);
+    }, [setRulerWidth, rulerWidthInPixels]);
+
+    /**
+     * Due to the performance hit resulting from representing each second of a long
+     * media item with an HTML div element, we need to use <canvas> to represent the
+     * timeline ruler.  And, because each browser has its own width limit for <canvas>
+     * we set the maximum canvas width at 8000px.  As a result, for longer timelines,
+     * we need to display the timeline ruler using one or more canvases.
+     */
+    const numberOfWholeCanvases = Math.floor(rulerWidthInPixels / MAX_CANVAS_WIDTH_IN_PIXELS);
+
+    const ticksPerWholeCanvas = MAX_CANVAS_WIDTH_IN_PIXELS / rulerUnitWidthInPixels;
+
+    const lastCanvasWidthInPixels = rulerWidthInPixels % MAX_CANVAS_WIDTH_IN_PIXELS;
+
+    const ticksInLastCanvas = lastCanvasWidthInPixels / rulerUnitWidthInPixels;
+
+    const numberOfCanvases = getNumberOfCanvases(lastCanvasWidthInPixels, numberOfWholeCanvases);
+
+    const numberOfCanvasesAsArray = [...Array(numberOfCanvases).keys()];
+
+    const canvases = numberOfCanvasesAsArray.reduce((acc: CanvasProps[], currentCanvas) => {
```

Comment:
This is quite readable, but also very complex. We should do some performance testing on longer audio files.